src/phocount.c

```diff
@@ -0,0 +1,189 @@
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT    
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS    
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE       
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,           
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES   
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR   
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)   
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,  
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OTHERWISE) ARISING   
+ * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE   
+ * POSSIBILITY OF SUCH DAMAGE.                                          
+ *
+ */
+
+#include <omp.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <math.h>
+#include <stdint.h>
+
+#include "phocount.h"
+
+int count(data_t *in, data_t *out, data_t *stddev, 
+          int ndims, index_t *dims, 
+          data_t *thresh, data_t *mean_filter, int sum_max, int nan){
+  index_t nimages = dims[0];
+  index_t M = dims[ndims-1];
+  index_t N = dims[ndims-2];
+  index_t imsize = N*M;
+
+  data_t nodata;
+  if(nan){
+    // Pad no data with nan not zero
+    nodata = NAN;
+  } else {
+    nodata = 0.0;
+  }
+
+  int x;
+  for(x=1;x<(ndims-2);x++){
+    nimages = nimages * dims[x];
+  }   
+
+  index_t i;
+#pragma omp parallel for shared(in, out, stddev) private(i)
+  for(i=0;i<nimages;i++){
+    // Find the start pointers of the image
+    data_t *inp = in + (i*imsize) - 1;
+    data_t *outp = out + (i*imsize) - 1;
+    data_t *stddevp = stddev + (i*imsize) - 1;
+
+    data_t pixel[9];
+
+    // Clear out the parts of the output array we don't use
+    
+    index_t j,k;
+    for(j=0;j<(M+1);j++){
+      inp++;
+      outp++;
+      stddevp++;
+      *outp = nodata;
+      *stddevp = nodata;
+    }
+
+    // Now start the search
+    for(j=1;j<(N-1);j++){
+      for(k=1;k<(M-1);k++){
+        inp++;
+        outp++;
+        stddevp++;
+
+        *outp = nodata;
+        *stddevp = nodata;
+
+        if((*inp < thresh[0]) || (*inp >= thresh[1])){
+          continue;
+        }
+
+        // The pixel is above thresh
+        // Now get the surrounding 9 pixels. 
+        pixel[0] = *inp;
+        pixel[1] = *(inp - M - 1);
+        pixel[2] = *(inp - M);
+        pixel[3] = *(inp - M + 1);
+        pixel[4] = *(inp - 1);
+        pixel[5] = *(inp + 1);
+        pixel[6] = *(inp + M - 1);
+        pixel[7] = *(inp + M);
+        pixel[8] = *(inp + M + 1);
+
+        // Is this the brightest pixel?
+        
+        int n;
+        int flag = 0;
+        for(n=1;n<9;n++){
+          if(pixel[n] > pixel[0]){
+            flag = 1;
+            break;
+          }
+        }
+
+        if(flag){
+          continue;
+        }
+      
+        // Sort the array
+        sort(pixel, 9);
```

Comment:
You sort the 9 pixels in intensity and then take the n most intense (such as the 3 most intense) That way you can tune how many pixels around a seed pixel you sum to get the photon energy